src/Features/Core/Portable/Diagnostics/EngineV2/DiagnosticIncrementalAnalyzer.Executor.cs

```diff
@@ -51,6 +58,41 @@ internal partial class DiagnosticIncrementalAnalyzer
             {
                 throw ExceptionUtilities.Unreachable;
             }
+
+            static bool IsAnalyzerSuppressedForDocument(
+                DiagnosticAnalyzer analyzer,
+                BackgroundAnalysisScope analysisScope,
+                bool isActiveDocument,
+                bool isOpenDocument,
+                bool isGeneratedRazorDocument)
+            {
+                if (isGeneratedRazorDocument)
+                {
+                    // This is a generated Razor document, and they always want all analyzer diagnostics.
```

Comment:
IIRC this exists because generated razor documents for open razor files are not actually 'open' in the workspace.  A user facing option here wouldn't be relevant I don't think.

The alternative solution here I think would be to mark the generated razor document as open, though I don't know what the implications of that are since there is no view ever associated with them.

Reply:
> IIRC this exists because generated razor documents for open razor files are not actually 'open' in the workspace. A user facing option here wouldn't be relevant I don't think.

I feel like i've heard about this now in at least 3 different places.  It feels like razor is broken here by not abiding by the concepts in the roslyn workspace, and so features then have to hack around this.

>though I don't know what the implications of that are since there is no view ever associated with them.

`IsOpen` doesn't really imply a `view`, merely that hte user is interacting with that doc directly in some fashion.  In razor that is true, even if it's because the contents of the razor doc are projected through teh combined file.  But conceptually that doc *is open* and being interacted with by the user. 

I feel ike we should just front load razor fixing this, and then we can remove all our special casing here.

Reply:
I wonder if in LSP razor scenarios we could just trigger Workspace.OpenDocument when we hear about the Razor generated document open.

The other alternative is for that to come through IDynamicFileInfoProvider (https://sourceroslyn.io/#Microsoft.CodeAnalysis.Workspaces/Workspace/Host/SourceFiles/IDynamicFileInfoProvider.cs,28), but doing it through LSP ensures it is all ordered correctly with other diagnostics requests.

Reply:
@jasonmalinowski  thoughts on the best place to insert this information. 

IIRC, IDynamicFileInfoProvider is how razor exposes documents to us, right?  So wouldn't it be fine for IsOpen to just see if a file is a dynamic file, and then just always say it's open (as the dynamic file only comes into existence when razor opens it, right)?

Reply:
IDynamicFileInfoProvider is also used for closed files; in some cases though when the razor file would switch from open to closed they'd empty out one file in favor of another -- i.e. the closed file may always be there, but it's just empty if the file is open and a different document might be used. That predates LSP though, so not sure if other things are happening now.

@NTaylorMullen can you clarify how this is all working again? :smile:

And so it's a bit unclear to me what this code is actually doing and why it's being treated as special; is there a summary of the Roslyn behavior here?

Reply:
> @NTaylorMullen can you clarify how this is all working again? 😄

You had it right. The IDynamicFileInfoProvider infra today serves two purposes:

1. Populate background C# content for Razor files
2. Allow open LSP documents to be fed into the C# workspace / control options (i.e. use Razor's tab vs. spaces options and not C#'s for its backing C# file)

Reply:
I have filed https://github.com/dotnet/roslyn/issues/58943 to track this cleanup work in a follow-up change.